Fix #3134: Make regex.Matcher report correct group positions.

MikaelMayer opened the pull request:
This is the first commit presenting a workaround for the absence of positions of matched groups in javascript regular expressions.
The workaround is total, meaning it works for any Javascript regular expression.
Under the hood, it parses the regular expression "manually", rebuilds a version where all sub-regex are matching groups, and can retrieve the position of each group by comparing their length.
The tests are passing, and I added a new test. for a typical case.

This fixes https://github.com/scala-js/scala-js/issues/3134

Before merging, I would like this code to be less messy, and add more tests to make sure it parses all kinds of regular expressions. Hence

* Where should I add those tests ?
* Is there a test suite in scalajs testing all regular expressions features ?
* How should I refactor the code? Is the lazy val enough?

scala-jenkins commented:
Can one of the admins verify this patch?

sjrd commented:
Dear Jenkins, test this please

sjrd commented:
> Where should I add those tests ?
> Is there a test suite in scalajs testing all regular expressions features ?

`RegexMatcherTest.scala` is the right place for this.

> How should I refactor the code? Is the lazy val enough?

IMO `parseRegex` and `Node` should definitely be in a separate `private[regex] object RegexParser`.

See also our [coding style guide](https://github.com/scala-js/scala-js/blob/master/CODINGSTYLE.md).

sjrd commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -175,15 +175,259 @@ final class Matcher private[regex] (
   def end(): Int = start() + group().length
   def group(): String = ensureLastMatch(0).get
 
+  lazy val startOf: (/*Group number*/Int => /*Start position*/Int) = {
```
`private`.

MikaelMayer replied:
Done

sjrd commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -202,7 +446,7 @@ final class Matcher private[regex] (
   // Seal the state
 
-  def toMatchResult(): MatchResult = new SealedResult(inputstr, lastMatch)
+  def toMatchResult(): MatchResult = new SealedResult(inputstr, lastMatch, startOf)
```
I do not think this is correct. `startOf` depends on `ensureLastMath`, and therefore it must be replicated in `SealedResult`, which has its own `lastMatch`.

MikaelMayer replied:
Oh I see ! You mean we might do something like "continue to next match" on the matcher, but if at this point we already called startOf, it's evaluated and we cannot modify it anymore, right?
So even the code in the Matcher is wrong actually.

MikaelMayer replied:
Let's just replace the lazy val by a def in Matcher. Then it becomes correct.

MikaelMayer replied:
I did something more advanced so that it works fine. But now there is no caching if it is not a sealed match result. Should I add caching in the original class ?

sjrd commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -175,15 +175,259 @@ final class Matcher private[regex] (
+      var start = 0
+      
+      // Takes a new group index and returns a new group index.
+      def setNewGroup(newGroupIndex: Int): Int = {
```
That description is not clear. Maybe

> Assigns consecutive group numbers starting from `newGroupIndex` to the nodes in this subtree, in a pre-order walk.
> Returns 1 plus the largest assigned group number.

MikaelMayer replied:
Taken into account.

sjrd commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -175,15 +175,259 @@ final class Matcher private[regex] (
+        regex match {
+          case Left(r) => newGroupIndex + 1
+          case Right(children) =>
+            ((newGroupIndex + 1) /: children) {
```
We don't use `/:`. Please use `foldLeft` instead.

sjrd commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -175,15 +175,259 @@ final class Matcher private[regex] (
+      }
+    
+      // First we record the matches
+      def setMatch(mr: Int => String): Unit = {
```
What does `mr` mean? Consider writing it in full. In general, avoid abbreviations except for codebase-wide well-known conventions (e.g., `pos` for `position`).

MikaelMayer replied:
I renamed these variables and others.

sjrd commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -175,15 +175,259 @@ final class Matcher private[regex] (
+    /** @param originalGroup Either the 1-based index of the original group that this node encloses, or the modifiers which are present before and after (ex: ?, *?, (?=...), (?!..))
+      * @param 
+      */
+    case class Node(originalGroup: Either[Int, (String, String)], regex: Either[String, Seq[Node]]) {
```
Those `Either`s seem weird. Why is this not an ADT with various `case class`es, rather than a single case class with `Either` fields?

MikaelMayer replied:
Replaced them by new more explicit classes with comments

```scala
// parentheses aspect of the regex
    sealed trait OriginalRegex {
      @inline def fold[A](numberF: Int => A, leftRightF: (String, String) => A): A = this match {
        case OriginallyGroupped(number) => numberF(number)
        case OriginallyWrapped(left, right) => leftRightF(left, right)
      }
    }
    // If the regex had originally a group number, that is, with simple parentheses
    // non-grouping parentheses (?: ...) are always added
    case class OriginallyGroupped(number: Int) extends OriginalRegex
    // If the regex was not originally groupped with parentheses,
    // then right holds the left and right modifiers (e.g. *, +, *?, +?, ?)
    case class OriginallyWrapped(left: String, right: String) extends OriginalRegex
    
    // Type of the node, a regex or a sequence of nodes.
    sealed trait NodeType {
      @inline def fold[A](regexF: String => A, childrenF: Seq[Node] => A): A = this match {
        case RegexLeaf(regex) => regexF(regex)
        case ParentNode(children) => childrenF(children)
      }
    }
    // A pure regex, without any subgroups
    case class RegexLeaf(regex: String) extends NodeType
    // A parent groupping, with children node. Modifiers to this group, such as parentheses, can be added using the field of Node originalGroup.
    case class ParentNode(children: Seq[Node]) extends NodeType
    
    /** @param originalGroup Either the 1-based index of the original group that this node encloses, or the modifiers which are present before and after (ex: ?, *?, (?=...), (?!..))
      * @param 
      */
    case class Node(originalGroup: OriginalRegex, nodeType: NodeType) {
```

sjrd commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -175,15 +175,259 @@ final class Matcher private[regex] (
+    
+      def end = start + (if(matched == null) 0 else matched.length)
+      
+      def buildOriginalRegex: String = originalGroup.fold(i => "(", s => "(?:" + s._1) + (regex match {
```
This is dead code.

MikaelMayer replied:
Removed.

sjrd commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -175,15 +175,259 @@ final class Matcher private[regex] (
+    @inline def finish(): (Int => Int) = {
+      val last = ensureLastMatch(0).toOption match {
+        case None => return (i: Int) =>
+          throw new Exception("regex did not match")
```
Avoid `return`. Instead, put the rest of the code in the `case Some`

MikaelMayer replied:
Done.

sjrd commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -175,15 +175,259 @@ final class Matcher private[regex] (
+                 if modifier == "?" || modifier == "*" || modifier == "+" || modifier == "*?"
+                  || modifier == "+?" || modifier.startsWith("{") =>
+                 (Node(Right(("", modifier)), Right(Seq(n))) +: tail, r, i)
+           case _ => res
```
What does this do? There's got to be a nicer way to write this stuff.

MikaelMayer replied:
If I parse a group (A) and then a modifier like the one mentioned, then  I rewrite the nodes by adding the modifiers as a right addition is the group was not matching, and wrap in another node if the group was already matching.

To be nicer, I pattern matched on the second node of the list, and then on the first.

sjrd commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -175,15 +175,259 @@ final class Matcher private[regex] (
+       def parse(s: String, nextGroupIndex: Int): (Seq[Node], String, Int)  = if (s.isEmpty) {
+         (Seq(), s, nextGroupIndex)
+       } else {
+         var res = s(0) match {
```
This match wants an `@switch` annotation.

MikaelMayer replied:
I pushed the `if`s inside the `case`s, added a `def` for the default case, and converted this pattern matching to constant pattern matching that now tolerates the `@switch` annotation

sjrd commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -175,15 +175,259 @@ final class Matcher private[regex] (
+         var res = s(0) match {
+         case '(' if s.length >= 3 && s(1) == '?' && (s(2) == '=' || s(2) == '!') =>
+           val (parsed, remaining, newNextGroupIndex) =
+             parse(s.substring(3), nextGroupIndex )
```
Taking `substring(x)` repeatedly causes the whole thing to be O(n²). Use an additional `index` parameter to `parse` instead.

MikaelMayer replied:
I removed almost all the calls to substring by adding an explicit parameter pIndex. Renamed s to "pattern", and removed "pattern" from the list of arguments since it is available locally.

sjrd commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -241,24 +485,27 @@ object Matcher {
         last(group).fold(-1) {
           groupStr => inputstr.indexOf(groupStr, last.index)
-        }
+        }*/
```
Commented out code.

MikaelMayer replied:
Removed

sjrd commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/regex/RegexMatcherTest.scala:
```diff
@@ -70,6 +70,40 @@ class RegexMatcherTest  {
   }
+  
+  @Test def start_end_group_tricky_and_toMatchResult(): Unit = {
+    val matcher = Pattern.compile("(Scala\\.js).*(Scala)$").matcher("Scala.js is a Scalable javascript compiler based on Scala")
```
Line length. (see coding style)

MikaelMayer replied:
Shortened

sjrd commented:
Also, you should rebase this on the 0.6.x branch, and edit your PR to target the 0.6.x branch.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4261/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4261/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5692/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5692/</a>
Test FAILed.

MikaelMayer commented:
I added some tests for (?! ...) and (?= ...) and made the program to work for those.
Also, the code of the method `parse` is refactored and much more readable.
Are there any constructs that we further need to test?

sjrd commented:
Dear Jenkins, retest this please

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4266/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4266/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5699/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5699/</a>
Test FAILed.

sjrd commented:
CI discovered a bunch of Scalastyle issues, which you can reproduce locally with

    > javalib/scalastyle

and/or

    > testSuite/test:scalastyle

as well as one failing partest: `run/ReplacementMatching.scala`. That one you can run locally with

    > partestSuite/testOnly -- --fastOpt run/ReplacementMatching.scala

MikaelMayer commented:
I just signed the CLA and am fixing these issues.

MikaelMayer commented:
i just fixed (hopefully the last) bugs after adding some coverage tests.
The traversal to propagate positions had to be done in the two directions: From the end to the start if we take the last match (*, +, {3,} ...), and from the start to the end if we just start a match test ( (?= ..,) (?! ...) ). For the remaining regexes, the start to the end is sufficient.

I also made the newly introduced methods private and refactored the test suite to apply the boy scout rule.

However, I could not reproduce the failing test because the command that you gave me finishes with the following exception

    sbt.ForkMain$ForkError: java.lang.RuntimeException: Error running C:\Users\[hidden]\Desktop\scala-js\partest\fetchedSources\2.11.11\test\files\run\ReplacementMatching.scala
    sbt.ForkMain$ForkError: java.io.IOException: Cannot run program "C:\Program Files\Java\jdk1.8.0_73\jre\bin\java.dll": CreateProcess error=193, %1 is not a valid Win32 application
    ForkMain$ForkError: java.io.IOException: CreateProcess error=193, %1 is not a valid Win32 application
    ...
    scala.tools.partest.scalajs.ScalaJSSuiteRunner$$anonfun$liftedTree1$1$1.apply(ScalaJSPartest.scala:102)

Can you please tell me what this error is about?

sjrd commented:
Jenkins, retest this please

sjrd commented:
I've never seen that error. It's very suspicious that `java` is a DLL. It should be an exe file. :-s

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4268/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4268/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5702/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5702/</a>
Test FAILed.

MikaelMayer commented:
Ok, what shall we do? Any log explaining the error?

Le mar. 12 sept. 2017 à 12:05, scala-jenkins <notifications@github.com> a
écrit :

> Refer to this link for build results (access rights to CI server needed):
>
> <https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4268/>
> https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4268/
> <https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5702/>
> https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5702/
> Test FAILed.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scala-js/scala-js/pull/3136#issuecomment-328806424>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADbytxQZnBOm9XLB_JAOkgnMDaTvNIopks5shld0gaJpZM4PRy3U>
> .
>

sjrd commented:
You can click on the links below, then go to Build Graph, and there you have a graph with all the jobs. The red ones are failed, and you can follow the console icon to see the logs.

MikaelMayer commented:
I found the failure, but it looks like I don't understand fully the regexes of Javascript.
See the following code, put it in ScalaFiddle:

```scala
val r2 = new js.RegExp("""(\d+)/(\d+)/(\d+)""")
r2.lastIndex = 44
println(r.lastIndex)
val m2 = r2.exec("1/1/2001 marks the start of the millennium. 31/12/2000 doesn't.")
println(m2)
```scala

I expected the print to display 31/12/2000, but it still matches the first 1/1/2001 instead. Why is that?

MikaelMayer commented:
Oh the w3c says that:

    Note: This property [lastIndex] only works if the "g" modifier is set.

MikaelMayer commented:
Ok I fixed the bug. After rewriting the regular expression, I was not carrying the flags anymore. I now carry over the flags g, m and i.

sjrd commented:
Dear Jenkins, retest this please

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4269/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4269/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5703/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5703/</a>
Test FAILed.

MikaelMayer commented:
Funny, the test just revealed a bug in the JVM regular expressions.
namely, in a JVM-based console,

    "ab(?:a(c))*ac".r.findFirstMatchIn("abacacac").get.start(1)

returns 7, as if the match on (c) was the last one.
This is obviously wrong. To make sure it is not Scala, I ran the same regular expression and string on a pure Java-based system here: http://www.regexplanet.com/advanced/java/index.html
and I obtain the same error.

In PHP, the same regex/string returns correctly 5, as for the implementation that I am submitting.
Shall I just remove the test if under the JVM ?

sjrd commented:
Are we sure this is a bug, as opposed to intended behavior of regex in Java? Is there an OpenJDK bug filed for it?

MikaelMayer commented:
It cannot be an intended behavior. `(?:...)` is a non-matching group but consumes the input. Since the regular expression ends with `ac`, and the string ends with `ac` at start position 6, the `c` in the non-matching group cannot have index 7.

I'm currently filling a bug report in http://bugreport.java.com/submit_intro.do . I could not find it anywhere. But bugs in regular expressions in the JVM are not new, see this SO question:
https://stackoverflow.com/questions/14712932/bug-in-java-regex-implementation

BTW the equivalent java code is the following:

    import java.util.regex.*;
    Pattern p = Pattern.compile("ab(?:a(c))*ac");
    Matcher m = p.matcher("abacacac");
    m.find();
    System.out.println(m.start(1));

sjrd commented:
Ah, I see. Yes, it has to be a bug. In that case, yes, you can use `if (!Platform.executingOnJVM)` or `assumeFalse("Bug on the JVM", Platform.executingOnJVM)` to exclude the test from the JVM test suite.

MikaelMayer commented:
Bug report number 9050822 filed.

MikaelMayer commented:
More precisely, it should be `executingInJVM`, not `executingOnJVM` (for subsequent readers of this thread)

MikaelMayer commented:
Wait, before merging, let me incorporate group matching semantics (I forgot them but they are important !)
For example, the following test will not work:

    parseExpect("bc(.c).c(\1)", "bczcxczc",    0, 8,  2, 4,  6, 8)

because the regular expression is translated to:

    (bc)(.c)(.c)(\1)

so the `\1` now refers to the wrong previous group.

MikaelMayer commented:
I just pushed some changes but I cannot test my new changes.
For some unknown reason, the repository refuses to build when I do the command line that I used to use since the beginning

   sbt "javalib/test-only *RegexMatcherTest"

it complains with the error:

```
Fast optimizing C:\Users\Mikael\Desktop\scala-js\javalib\target\scala-2.11\java-library-for-scalajs-test-fastopt.js
[error] Fatal error: java.lang.Object is missing
[error]   called from core module analyzer
org.scalajs.core.tools.linker.LinkingException: There were linking errors
```

Even when I stash my changes, reverting the repository back to a state where it used to work, it does not work anymore.
I redownloaded the repository, built from scratch, it worked once and then if tailed again. Hopefully the tests are now passing.

sjrd commented:
Dear Jenkins, ok to test

sjrd commented:
The exception would thrown if, for some reason, the classpath given to the linker does not contain the definition of `java.lang.Object`. Usually that's because it's empty. On Windows I have seen it happen if a process holds a lock on `scalajs-library.jar` while sbt tries to modify it, or things like that.

sjrd commented:
I will review this better next week. I have a paper deadline at the end of the week.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4274/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4274/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5709/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5709/</a>
Test FAILed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4275/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4275/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5710/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5710/</a>
Test FAILed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4276/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4276/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5711/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5711/</a>
Test FAILed.

MikaelMayer commented:
I published the bug here:
http://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8187537

MikaelMayer commented:
I added a (very lazy) cache to store the calculated group positions for a match at a given position. This cached is implemented as a lazy val for `MatchResult`.
Else I saw that for every call to start or end, it would recompute and rematch the regular expression.

I also fixed the last bug (missing backslash). remove trailing whitespace from lines in the test suite.
I'm good to merge this PR.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4277/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4277/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5719/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5719/</a>
Test FAILed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4278/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4278/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5720/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5720/</a>
Test FAILed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4279/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4279/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5721/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5721/</a>
Test FAILed.

sjrd commented:
Our CI is of the unforgiving kind ;) Looks like the only remaining things are a couple spaces in the tests:
```
[error] .../testsuite/javalib/util/regex/RegexMatcherTest.scala:97:6: Public method must have explicit type
[error] .../testsuite/javalib/util/regex/RegexMatcherTest.scala:147:24: Space before token ,
[error] .../testsuite/javalib/util/regex/RegexMatcherTest.scala:148:24: Space before token ,
[error] .../testsuite/javalib/util/regex/RegexMatcherTest.scala:156:24: Space before token ,
[error] .../testsuite/javalib/util/regex/RegexMatcherTest.scala:157:24: Space before token ,
```

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4280/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4280/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5722/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5722/</a>
Test FAILed.

sjrd commented:
Dear Jenkins, retest this please

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4281/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4281/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5723/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5723/</a>
Test PASSed.

MikaelMayer commented:
All checks have passed. Good to merge.

sjrd commented:
> Good to merge.

Well ... let me give it an additional review first ;)

sjrd commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/regex/RegexMatcherTest.scala:
```diff
@@ -33,42 +33,129 @@ class RegexMatcherTest  {
+    assertEquals(startEndMatch(0)._1, matcher.start)
+    assertEquals(startEndMatch(0)._2, matcher.end  )
+    assertEquals(startEndMatch(0)._3, matcher.group)
+    for (((start, end, mtch), i) <- startEndMatch.zipWithIndex) {
```
You should add
```scala
assertEquals(startEndMatch.size - 1, matcher.groupCount)
```
above this line, and then you can get rid of the `assertEquals(N, matcher.groupCount)` that are in all the specific test cases.

Unless I'm missing something ...

MikaelMayer replied:
Did so. Also, I factored the `assertTrue(matcher.find())` in `checkGroups` since they are always called before.

sjrd commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/regex/RegexMatcherTest.scala:
```diff
@@ -33,42 +33,129 @@ class RegexMatcherTest  {
+    )
+  }
 
+  def parseExpect(regex: String, str: String, pos: Int*): Unit = {
```
This would be much clearer if it took a `startsAndEnds: (Int, Int)*`. It would make the call sites much more readable as they would be written as `(0, 2), (1, 2)` instead of `0, 2, 1, 2`. Or even `0 -> 2, 1 -> 2`, which is arguably a nice DSL for `start -> end`.

MikaelMayer replied:
New DSL Start -> End implemented.

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -2,14 +2,14 @@ package java.util.regex
     private var pattern0: Pattern, private var input0: CharSequence,
     private var regionStart0: Int, private var regionEnd0: Int)
-    extends AnyRef with MatchResult {
+    extends AnyRef with MatchResult with StartHack {
```
Moved to a private lazy val startHack for Matcher, inlined the call for the SealedResult. This works too.

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -223,6 +219,21 @@ final class Matcher private[regex] (
+  private[regex] def getRegex(): js.RegExp = regexp
+
+  // For performance reasons, if we don't use the positions
+  private var cache: Map[Int, Int => Int] = Map()
```
Considered and renamed.

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -223,6 +219,21 @@ final class Matcher private[regex] (
+      val groupPositions = computeCurrentGroupPositions()
+      cache += start() -> groupPositions
+      groupPositions
+    })
```
I simplified the process, and removed the dependency to start() so that it invalidates the cache `Option[Int => Int]` every time there is a call to `find`, `reset` or `usePattern`
I also renamed StartHack to GroupStartMap.

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -241,24 +252,21 @@ object Matcher {
   private final class SealedResult(inputstr: String,
-      lastMatch: js.RegExp.ExecResult) extends MatchResult {
+      lastMatch: js.RegExp.ExecResult,
+      val pattern: Pattern) extends MatchResult with StartHack {
```
I did things consistently.

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -241,24 +252,21 @@ object Matcher {
     def end(): Int = start() + group().length
     def group(): String = ensureLastMatch(0).get
 
+    private lazy val fixedStartOf = computeCurrentGroupPositions()
```
Exactly. This is what I just did.

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -275,5 +283,358 @@ object Matcher {
+  }
+}
+
+private[regex] trait StartHack {
```
Moved to a new file. Added comments

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -275,5 +283,358 @@ object Matcher {
+
+  def start(): Int
+
+  private[regex] def getRegex(): js.RegExp
```
They did indeed.

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -275,5 +283,358 @@ object Matcher {
+  private[regex] def getRegex(): js.RegExp
+
+  private[regex] def computeCurrentGroupPositions(): (Int => Int) = {
+    val regex = getRegex
```
I also use other properties of regex: so I'm afraid I cannot avoid that.

```
{
      (if (regex.global) "g" else "") +
      (if (regex.ignoreCase) "i" else "") +
      (if (regex.multiline) "m" else "")
    }
```

MikaelMayer replied:
I switched to use the Pattern instead. Good remark. I recover the flags using >> and &, hopefully the flag API should not change.
I wonder however if always setting the flag global to true is the thing to do, since there is no equivalent to global for a pattern. (and Pattern.compile always uses the "g" flag)

sjrd replied:
IIRC, Java's regexes always behave as JS *global* regexes, hence why `Pattern.compile` always uses the `g` flag.

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -275,5 +283,358 @@ object Matcher {
+    node.setNewGroup(1)
+    val mapping = node.obtainMapping
+    node.transformGroupNumber(mapping.mapValues(_.newGroup))
+    val allMatchingRegex_s = node.buildRegex
```
Renamed

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -275,5 +283,358 @@ object Matcher {
+  }
+
+  // parentheses aspect of the regex
+  private[regex] sealed trait OriginalRegex {
```
Moved traits and case classes to companion object. Put `private` instead of `private[regex]` where I could.

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -275,5 +283,358 @@ object Matcher {
+  }
+  // If the regex had originally a group number, that is, with simple parentheses
+  // non-grouping parentheses (?: ...) are always added
+  private[regex] case class OriginallyGroupped(number: Int) extends OriginalRegex
```
Fixed typo. Final case classes everywhere.

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -275,5 +283,358 @@ object Matcher {
+  // non-grouping parentheses (?: ...) are always added
+  private[regex] case class OriginallyGroupped(number: Int) extends OriginalRegex
+  // If the regex was not originally groupped with parentheses,
+  // then right holds the left and right modifiers (e.g. *, +, *?, +?, ?)
```
then left and right are modifiers

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -275,5 +283,358 @@ object Matcher {
+             after (ex: ?, *?, (?=...), (?!..))
+    * @param nodeType The type of the node (Regexleaf, or ParentNode)
+    */
+  private[regex] case class Node(originalGroup: OriginalRegex, var nodeType: NodeType) {
```
Added a comment. "Is a var because we need to shift the backreference groups".
Node is a case class for the sake of easy construction and debugging.

sjrd replied:
If it's only for easy construction and debugging, it is preferable to define a `def apply` in the companion object, and override `toString()`.

MikaelMayer replied:
Did so.

MikaelMayer replied:
Actually, I also need to deconstruct the Node for pattern matching. Is it still worth having a object with `apply/unapply` and the `toString` override? I thought that case classes were doing exactly this.

sjrd replied:
Hum, I see. Well, it's up to you. Whichever you prefer. Probably at this point the case class was a better idea, although I'd call it a code smell if we unapply mutable data structures. But since this is confined in a private area, we need not be that strict.

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -275,5 +283,358 @@ object Matcher {
+        case OriginallyWrapped.Absolute() => offset
+        case _ => end
+      }
+    }
```
That was the trickiest part I had to solve, when assigning group positions. I could not choose between assigning group numbers from left to right or from right to left, because there both failed in one case each. Normally, both directions give the same result. But there are corner cases.
Consider the following regex matching `abbbbbbc`

    (?=ab*(c))ab

After conversion, this becomes:

    (?=(ab*)(c))(ab)

To recover the position of the group (c), we cannot do anything but compute it from the length of (ab*), that is, propagate the start, compute the length, and return the end, and this, recursively. This is what we need to do in a forward-matching regex.

However, consider the following regex matching `abcbdbe`

    a(b.)*

After conversion, it is transformed to:

    (a)((b.)*)

The semantics of group matching under a star are that the last match is considered
Therefore, we cannot obtain the start position of (b.) from propagating lengths from left to right..
What we first do is to get the start, then the end, of the group `((b.)*)`, and then we propagate this end to the inner group.

Note that when javascript will support back-matching `(?<= )` and `(?<! )` (hopefully one day), we will be able to implement the length transfer using the `setEndReturnStart` method, because we have no clue on where the match started (this is different from the `start` position because it can extend before it)

`Absolute()` means a marker of the type `(?= )` or `(?! )`, they do not count to propagate the length on the right or on the left.

`Repeater()` designates all the repeat-transformer which have more or less the semantics of *. Every group having a repeater uses teh semantics of propagating the end to the start.

sjrd replied:
OK, thanks. Looks like you could copy-paste this as a big comment in the code ;)

MikaelMayer replied:
Pasted in the code as a big comment.

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -275,5 +283,358 @@ object Matcher {
+      originalGroup.fold(groupNum => "(", (left: String, right: String) => "((?:" + left) +
+      nodeType.     fold(regex => regex,
+        children => "(?:" + children.map(_.buildRegex).mkString("") + ")") +
+      originalGroup.fold(groupNum => ")", (left: String, right: String) => ")" + right + ")")
```
Made the code cleaner to read. Inserted three `val`s.

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -275,5 +283,358 @@ object Matcher {
+        children => "(?:" + children.map(_.buildRegex).mkString("") + ")") +
+      originalGroup.fold(groupNum => ")", (left: String, right: String) => ")" + right + ")")
+
+    def obtainMapping: Map[Int, Node] = {
```
I renamed it to `getGroupNodeMap` and the associated variable to `groupNodeMap` since it is a mapping from original group numbers to their associated node. From this mapping, we can retrieve both the new group so that we can shift backreferences, and we can also retrieve the wanted mapping group number to start.

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -275,5 +283,358 @@ object Matcher {
+
+  // Wraps every char in the regexp in a group to find the intermediate positions.
+  // Same for subgroups.
+  // (?:AB)C => (?:(AB))(?:(C))
```
I always insert the `(?: )` to make sure the repeating modifiers apply on the entire regular expression. Indeed, consider the following regex:

    ab*c

This one I transform it to

    (a)((?:b)*)(c)

so that the `(?:` operator is transparent., and it is not different if there are multiple chars under the star.

I fixed the doc. Originally, I was matching every single char. But later I realised that I could simplify the regex by merging chars if they were not part of different original groups.

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -275,5 +283,358 @@ object Matcher {
+  // (?:AB)C => (?:(AB))(?:(C))
+  // Transform:  (A(B))C(D)  to  (((A)(B))(C)(D)) so that using the length
+  // of the 1st group and the 3rd, we can obtain the start position of D.
+  private[regex] def parseRegex(pattern: String): Node = {
```
The two steps are done at the same time for performance reason. You could build the original regex using the (now deleted on your request) `buildOriginalRegex` function. So it's indeed a parse.
But the parse is done so that there is enough information (original group, repeaters, children...) to build the new regular expression from another kind of traversal of the node.

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -275,5 +283,358 @@ object Matcher {
+     // Take care of escaped parentheses \( and \)
+     // Takes care of non-group parenthesis (?:)
+     // Returns a sequence of nodes, the remaining non-parsed, and the next group index
+     def parseClosingParenthesis(pIndex: Int) = pIndex + 1
```
Indeed. Moved.

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -275,5 +283,358 @@ object Matcher {
+          case Node(OriginallyWrapped("", ""), RegexLeaf(r1)) +:
+               Node(OriginallyWrapped("", ""), RegexLeaf(r2)) +: tail
+               if ! (r1.length >= 2 && r1(0) == '\\' && r1(1).isDigit) &&
+                  ! (r2.length >= 2 && r2(0) == '\\' && r2(1).isDigit)               =>
```
How do you negate such an extractor? Here i want to merge regular expressions iff none of them is a back-reference.

MikaelMayer replied:
I refactor the method to test for back-references.

MikaelMayer replied:
and thanks for the code of the object.

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -275,5 +283,358 @@ object Matcher {
+                  pIndex + 2,
+                  nextGroupIndex
+                )
+              }
```
Thanks for the code. I used your patch and it works.

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -275,5 +283,358 @@ object Matcher {
+              addSiblings(
+                Node(OriginallyWrapped("", pattern(pIndex) + ""), RegexLeaf("")),
+                pIndex + 1, nextGroupIndex)
+            }
```
Thanks for the code. I need to review it because after inserting it, the tests failed.

sjrd replied:
Oh yes, on the last line `pIndex + 2` should be `nextIndex`.

MikaelMayer replied:
That made the tests pass. Thanks.

MikaelMayer commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -275,5 +283,358 @@ object Matcher {
+
+    parse(0, 1)._1 match {
+       case Nil => Node(OriginallyWrapped("",""), RegexLeaf(""))
+       case Seq(n:Node) => n
```
Removed `: Node`

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4282/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4282/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5726/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5726/</a>
Test PASSed.

MikaelMayer commented:
LGTM

Le jeu. 21 sept. 2017 à 18:04, scala-jenkins <notifications@github.com> a
écrit :

> Refer to this link for build results (access rights to CI server needed):
>
> <https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4282/>
> https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4282/
> <https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5726/>
> https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5726/
> Test PASSed.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scala-js/scala-js/pull/3136#issuecomment-331203375>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADbyt3p_G976v4RifW4uiWU8v5cnOWmvks5skoj1gaJpZM4PRy3U>
> .
>

sjrd commented:
If you don't mind, tomorrow or this week-end I will push a commit to fix the coding style to your branch. Then you can squash all the commits into one (or I can do it), and it should be good to go.

MikaelMayer commented:
Today I'm not in office, you can do whatever you want for the coding style
I can squash the commits too..

Le ven. 22 sept. 2017 à 00:43, Sébastien Doeraene <notifications@github.com>
a écrit :

> If you don't mind, tomorrow or this week-end I will push a commit to fix
> the coding style to your branch. Then you can squash all the commits into
> one (or I can do it), and it should be good to go.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scala-js/scala-js/pull/3136#issuecomment-331301929>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADbyt9nx85YCJg4AH-zR5DsTjctI2kVLks5skuZ7gaJpZM4PRy3U>
> .
>

MikaelMayer commented:
I just figured out that I forgot to test the engine with disjunctions (the | operator). I don't know how I completely forgot about it !

And it looks like it is currently not working with basic tests (such as 'a|(b)')  I need to fix that.

MikaelMayer commented:
Ok this is now fixed and tested.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4288/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4288/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5747/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5747/</a>
Test FAILed.

MikaelMayer commented:
I don't understand. The new tests is passing on my machine, not on jenkins. Impossible to debug that.

sjrd commented:
I'll have a look this afternoon.

sjrd commented:
Ouch. They also pass on my machine :-s
This is not going to be fun. I'll keep looking.

sjrd commented:
I expected and hoped it was a Node.js version discrepancy issue, but I invalidated that. The CI uses Node.js 6.9.5. Trying with that exact version on my machine does not make the tests fail :-(

sjrd commented:
Oh! The tests pass *on JS*. They fail on the JVM! You can reproduce with

    > testSuiteJVM/test

in sbt.

The CI log actually says that, if you Ctrl+F for "RETRYING". It is what I do first when I read the logs, to see exactly what command failed, but apparently I glanced over the "JVM" part of the command line :s

MikaelMayer commented:
Ok, after investigation, this is another difference of regex handling between java and javascript (as tested on a regex engine like regex101.com)

The match of groups inside a repeated disjunction

* are the match of the last time this disjunction matched in the JVM
* are the match of the last occurrence of this disjunction in javascript

Therefore, even the *groups* are not extracted the same. For example:

    ^(a(.)|b(.))+$

after pattern matching on "axby"

* In Javascript the groups are "axby", "by", N/A, "y"
* In Javalib the groups are  "axby", "by", "x", "y"

I don't know which extraction is standard.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4289/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4289/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5749/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5749/</a>
Test PASSed.

sjrd commented:
Ran out of time to reformat the two other files today; I'm catching a plane. Will finish on Monday.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4291/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4291/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5752/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5752/</a>
Test PASSed.

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4300/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4300/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5765/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5765/</a>
Test PASSed.

sjrd commented:
OK, I think this is now good for a giant squash, and then merge. To abide by our commit message guidelines, the first line should of the form:

    Fix #3134: Do something.

I suggest:

    Fix #3134: Make regex.Matcher report correct group positions.

The rest of the commit message (after a blank line) should give a short overview of how it is done. The first paragraph of the leading comment for `GroupStartMap` seems like a good candidate.

MikaelMayer commented:
I will do it tomorrow.
Thanks !!

Le lun. 9 oct. 2017 15:55, Sébastien Doeraene <notifications@github.com> a
écrit :

> OK, I think this is now good for a giant squash, and then merge. To abide
> by our commit message guidelines, the first line should of the form:
>
> Fix #3134: Do something.
>
> I suggest:
>
> Fix #3134: Make regex.Matcher report correct group positions.
>
> The rest of the commit message (after a blank line) should give a short
> overview of how it is done. The first paragraph of the leading comment for
> GroupStartMap seems like a good candidate.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scala-js/scala-js/pull/3136#issuecomment-335165357>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADbyt0OqcLmadvi-TTMlYeCrOPkLfKCHks5sqiW7gaJpZM4PRy3U>
> .
>

scala-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 

<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4302/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-pr/4302/</a>
<a href='https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5768/'>https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/5768/</a>
Test PASSed.
